Plans: Add expired state to `PlanStatus`

drewblaisdell opened the pull request:
Addresses part of #1471.

![screen shot 2015-12-17 at 5 23 38 pm](https://cloud.githubusercontent.com/assets/1130674/11886861/919fec54-a4e3-11e5-9873-460a1fab7a17.png)

**Testing**
- Visit http://calypso.localhost:3000/plans/:site for a site with a trial that is in the grace period (within three days of expiring).
- If there are three days left, assert that the notice reads 'EXPIRED TODAY', and that the remaining time text reads 'Plan features will be removed in 3 days'.
- If there are two days left, assert that the notice reads 'EXPIRED 1 DAY AGO', and that the remaining time text reads 'Plan features will be removed in 2 days'.
- If there is one day left, assert that the notice reads 'EXPIRED 2 DAYS AGO', and that the remaining time text reads 'Plan features will be removed in 1 day'.
- If it is the day of the expiry, assert that the notice reads 'EXPIRED 3 DAYS AGO', and that the remaining time text reads 'Plan features will be removed momentarily'.

**Reviews**
- [x] Code
- [x] Product

drewblaisdell commented on client/lib/site-specific-plans-details-list/assembler.js:
```diff
@@ -7,7 +7,7 @@ function createSiteSpecificPlanObject( plan ) {
 		currentPlan: Boolean( plan.current_plan ),
 		expiry: plan.expiry,
-		expiryMoment: moment( plan.expiry ),
+		expiryMoment: moment( plan.expiry ).startOf( 'day' ),
```
@Tug The issue we encountered during the meetup is simplified a little bit by:

- Ensuring that **all** moments (those in this assembler and the current moment calculated with `moment()`) are set to the start of the day.
- Using `moment().startOf( 'day' )` instead of `moment().set( { hour: 0, ... } )` each time.

mtias commented on client/my-sites/plans/plan-overview/plan-status/style.scss:
```diff
@@ -25,6 +25,10 @@
 	}
+
+	.notice {
+		text-transform: uppercase;
```
This should be lowercase given the new notices design.

drewblaisdell replied:
I'll update it. cc'ing @mikeshelton1503 so he sees this.

mtias commented on client/my-sites/plans/plan-overview/plan-status/index.jsx:
```diff
@@ -41,7 +65,9 @@ const PlanStatus = React.createClass( {
 				<CompactCard className="plan-status__info">
-					<div className={ iconClasses } />
+					<div className={ iconClasses }>
+						{ isInGracePeriod( plan ) && <Gridicon icon="notice" /> }
```
It'd be nice to add this to the `SiteIndicator` component as another case.

drewblaisdell replied:
Good idea. The data we need here to determine if a site is in the grace period isn't returned by the `/me/sites` endpoint, so this will take some thought. I opened an issue for it here: #1848

drewblaisdell commented:
This has been rebased to pull in changes from #1708 and is ready for review.

stephanethomas commented on client/lib/plans/index.js:
```diff
@@ -3,8 +3,32 @@
+		return true;
+	}
+
+	return false;
```
What about the following simpler approach?

```javascript
export function isInGracePeriod( plan ) {
 return ( getDaysUntilUserFacingExpiry( plan ) <= 0 ) && ( getDaysUntilExpiry( plan ) >= 0 );
}
```

stephanethomas commented on client/my-sites/plans/plan-overview/plan-status/index.jsx:
```diff
@@ -31,6 +33,28 @@ const PlanStatus = React.createClass( {
+					? this.translate( 'Expired today' )
+					: this.translate(
+						'Expired %(daysAfterUserFacingExpiry)d day ago',
+						'Expired %(daysAfterUserFacingExpiry)d days ago', {
```
We shouldn't probably use `daysAfterUserFacingExpiry` here since translators would not know what that is (I'm not sure either). What about just `daysAfterExpiry` or just `days` instead?

stephanethomas commented on client/my-sites/plans/plan-overview/plan-status/index.jsx:
```diff
@@ -31,6 +33,28 @@ const PlanStatus = React.createClass( {
+						'Expired %(daysAfterUserFacingExpiry)d days ago', {
+							args: { daysAfterUserFacingExpiry },
+							count: daysAfterUserFacingExpiry
+						} );
```
That might be just personal preference but I find this indented declaration difficult to read. What do you think of the following?

```javascript
const daysAfterUserFacingExpiry = Math.abs( getDaysUntilUserFacingExpiry( plan ) );
let noticeText;

if ( daysAfterUserFacingExpiry === 0 ) {
  noticeText = this.translate( 'Expired today' );
} else {
  noticeText = this.translate(
  'Expired %(daysAfterUserFacingExpiry)d day ago',
  'Expired %(daysAfterUserFacingExpiry)d days ago', {
    args: { daysAfterUserFacingExpiry },
    count: daysAfterUserFacingExpiry
  } );
}

return (
  <Notice isCompact status="is-error">
    { noticeText }
  </Notice>
);
```

Since we're not using `noticeText` more than once we could also imagine this shorter approach:

```javascript
const daysAfterUserFacingExpiry = Math.abs( getDaysUntilUserFacingExpiry( plan ) );

return (
  <Notice isCompact status="is-error">
    { daysAfterUserFacingExpiry === 0
      ? this.translate( 'Expired today' )
      : this.translate(
      'Expired %(daysAfterUserFacingExpiry)d day ago',
      'Expired %(daysAfterUserFacingExpiry)d days ago', {
        args: { daysAfterUserFacingExpiry },
        count: daysAfterUserFacingExpiry
      } ) }
  </Notice>
);
```

stephanethomas commented on client/my-sites/plans/plan-overview/plan-status/index.jsx:
```diff
@@ -55,12 +81,14 @@ const PlanStatus = React.createClass( {
 						className="plan-status__button"
-						onClick={ this.purchasePlan }>
+						onClick={ this.purchasePlan }
+					>
```
Did you return to the line on purpose? In that case we should probably discuss it internally to make sure everybody is on the same page as I don't remember seeing this approach very often in our code base.

drewblaisdell replied:
Yes, this is a new rule in our [eslint config](https://github.com/Automattic/wp-calypso/blob/master/.eslintrc#L61).

stephanethomas replied:
Gotcha, thanks!

stephanethomas commented on client/my-sites/plans/plan-overview/plan-status/progress.jsx:
```diff
@@ -17,22 +17,41 @@ const PlanStatusProgress = React.createClass( {
 				value={ progress }
-				total={ trialPeriodInDays } />
+				total={ trialPeriodInDays }
+			/>
```
Same thing here :).

stephanethomas commented:
I guess this shouldn't happen (well, I'm not quite sure about that) but I've noticed the following display bug once the grace period is over:

![screenshot](https://cloud.githubusercontent.com/assets/594356/11915225/dfe79028-a69c-11e5-94f1-229df1ccd84e.png)

stephanethomas commented:
I added a few remarks but otherwise this works as advertised.

drewblaisdell commented:
Thanks for looking at this, @stephanethomas. I updated the commits to address your feedback with the following changes:

- Updated the syntax of `isInGracePeriod`.
- Switched to use `days` instead of `daysUntilUserFacingExpiry` in `renderNotice`.
- Updated the syntax of the notice text definition in `renderNotice`.
- Updated `PlanStatusProgress` to handle the case where the trial was not removed. Ideally, users shouldn't ever be in this state, but it is easy enough to handle.

spncrb commented:
LGTM - only concern is the styling.  
Even with a max-viewport window, the text takes up two lines:
![screen shot 2015-12-21 at 4 58 33 pm](https://cloud.githubusercontent.com/assets/8921630/11945403/67e1eef2-a804-11e5-9e2f-e054350b993f.png)
It looks a lot better in the mockup above, with the text on a single line (WordPress.com Premium Free Trial)

Tug commented:
Some margin must have been added or something, so it wasn't fitting into one line in english.
I fixed that by decreasing the font size a little bit

Also added some comment and reviewed the code :+1: LGTM

stephanethomas commented:
I really don't think we should change the size of the font to avoid having two lines of text. I bet it's longer in French and in other languages. If this is really an issue, we should prevent it from wrapping using the appropriate CSS property and take advantage of the flexbox layout to move the button below if there is no space left.

mikeshelton1503 commented:
For the font-size, the original intended size was `21px` so that should fix the English version at max-width. Sorry I should have indicated that originally.

For other languages it may still wrap but I'm ok with this. Stephane's suggestion could work too but we're not gaining too much with that method because the overall height of the section would be taller so it would just be a trade off of evils.

drewblaisdell commented:
I updated @Tug's commit to change the font size to `21px`. :+1:
